File: Models/AppState.cs
using System;
using System.Collections.Generic;
using System.Text.Json.Serialization;

namespace AnimeWallpaper.Models;

public class AppState
{
    [JsonPropertyName("usedWallpaperIds")]
    public List<string> UsedWallpaperIds { get; set; } = new();

    [JsonPropertyName("currentWallpaperId")]
    public string CurrentWallpaperId { get; set; } = string.Empty;

    [JsonPropertyName("currentWallpaperPath")]
    public string CurrentWallpaperPath { get; set; } = string.Empty;

    [JsonPropertyName("currentResolution")]
    public string CurrentResolution { get; set; } = string.Empty;

    [JsonPropertyName("lastChanged")]
    public DateTime? LastChanged { get; set; }

    [JsonPropertyName("isPaused")]
    public bool IsPaused { get; set; } = false;

    public void AddUsedId(string id, int maxTracked = 200)
    {
        if (string.IsNullOrWhiteSpace(id)) return;

        UsedWallpaperIds.Remove(id);
        UsedWallpaperIds.Add(id);

        if (UsedWallpaperIds.Count > maxTracked)
        {
            UsedWallpaperIds.RemoveRange(0, UsedWallpaperIds.Count - maxTracked);
        }
    }
}

File: macos/Sources/Models/AppState.swift
import Foundation

public struct AppState: Codable {
    public var usedWallpaperIds: [String]
    public var currentWallpaperId: String
    public var currentWallpaperPath: String
    public var currentResolution: String
    public var lastChanged: Date?
    public var isPaused: Bool

    public init(
        usedWallpaperIds: [String] = [],
        currentWallpaperId: String = "",
        currentWallpaperPath: String = "",
        currentResolution: String = "",
        lastChanged: Date? = nil,
        isPaused: Bool = false
    ) {
        self.usedWallpaperIds = usedWallpaperIds
        self.currentWallpaperId = currentWallpaperId
        self.currentWallpaperPath = currentWallpaperPath
        self.currentResolution = currentResolution
        self.lastChanged = lastChanged
        self.isPaused = isPaused
    }

    public mutating func addUsedId(_ id: String, maxTracked: Int = 200) {
        guard !id.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty else { return }
        usedWallpaperIds.removeAll { $0 == id }
        usedWallpaperIds.append(id)
        if usedWallpaperIds.count > maxTracked {
            usedWallpaperIds.removeFirst(usedWallpaperIds.count - maxTracked)
        }
    }
}
